### DISCUSSION TITLE
Forcing certain vehicles to visit certain locations

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#2850] [Routing (and legacy CP) questions]
Hello,

Disclosure: I used the search, did not find much and what I found was kind of contradictory. So please allow me to ask this question.

I am solving a CVRPTW and it seems to be working OK. I have a fleet of vehicles to go around town and pickup passengers and specified locations and drop them off at other locations at constrained pickup and drop-off times. Sometimes there's a requirement that certain passenger (location) is visited only by a predetermined vehicle. Meaning that the or-tools should find optional solution for particular vehicle to go around town for other pickups and drop-offs, but it has to be there for particular pickup (and thus drop-off). How should I go around that? Should I explore the VehicleVar route or use ApplyLocks or ApplyLocksToAllVehicles? I've used

```cs
RoutingModel.VehicleVar(pickupLocationIndex).SetValues(new long[] { -1, vehicleIndex});
RoutingModel.VehicleVar(dropoffLocationIndex).SetValues(new long[] { -1, vehicleIndex});
```
but the solutions drops already pre-assigned locations. **In my case It should rather not find a solution at all rather than drop a predefined pickup.**

Any insight would be appreciated.

Thanks,
Oleg D.

Ref:
https://github.com/google/or-tools/blob/86d4c543f717fa8716a9d66c25143660668bf825/ortools/constraint_solver/routing.h#L1141-L1159

**Comment by odashevsky** [comment]
Hi,
Kindly, at least, I was wondering how to specify "empty" locks in call to ApplyLocksToAllVehicles()? I mean, the call takes long[][]. How do I specify in that parameter something like this (for 3 vehicles):
`long[][] locks = new long[][] { {}, {null,null, 3,4,null, null}, {1,2,null,null,null,null} }`
meaning 'node 3 and 4 should be visited by vehicle 1, node 1 and 2 should be visited by vehicle 2, How would I do that? 
Thanks.

**Reply by Mizux** [reply]
in your case I would use:

Looking at the swig generated C#
from my `<buildir>/dotnet/ortools/constraint_solver/RoutingModel.cs`
```cs
  public bool ApplyLocksToAllVehicles( long[][]  locks, bool close_routes) {
    bool ret = operations_research_constraint_solverPINVOKE.RoutingModel_ApplyLocksToAllVehicles(swigCPtr, 
  locks.GetLength(0),
  NestedArrayHelper.GetArraySecondSize(locks),
  NestedArrayHelper.GetFlatArray(locks)
, close_routes);
    return ret;
  }
```

So I would use:
```cs
one = manager.NodeToIndex(1);
two = manager.NodeToIndex(2);
three = manager.NodeToIndex(3);
four = manager.NodeToIndex(4);
long[][] locks = new long[][] { {}, {three,four}, {one, two} }
routing.ApplyLocksToAllVehicles(locks, false);
```

Please note lock freeze the chain `A -> B` while you want `A -> * -> B` or `B -> * -> A`, aka on the same route no necessary on  the same order with possible node in between...

**Reply by odashevsky** [reply]
Thanks for the reply. I tried that and the solver generates no solution, though I know there're vehicle w/ out locks and could make it to locations not involved in the locks. I'll keep digging, but anything I should watch out for?
Good point, some times I may need A -> * -> B, sometimes A-> B. I will try using locks for A-> B, but how do I go around A-> * -> B? Should I use
`AddSoftSameVehicleConstraint ` with `nodes = [A.B] `
or
`model.VehicleVar(A).SetValues(new long[] { vehicleIndex});` 
`model.VehicleVar(B).SetValues(new long[] { vehicleIndex});` 
?

**Reply by odashevsky** [reply]
@Mizux, 
Very strange: when I have a time constraint (pickup and delivery windows) and I put a lock on ONLY ONE vehicle (out of 32) to visit 3 nodes (out of 108), the solver quits immediately with no solution found. Without the lock, a good solution is found fairly quickly. What should I check or look into? Thanks.

**Reply by odashevsky** [reply]
Any ideas?

**Reply by Mizux** [reply]
> Should I use
 `AddSoftSameVehicleConstraint ` with `nodes = [A.B] `
 or
 `model.VehicleVar(A).SetValues(new long[] { vehicleIndex});`
 `model.VehicleVar(B).SetValues(new long[] { vehicleIndex});` ?

The former only add penalty if both node are not on the same vehicle whatever the vehicle is.
The latter force (hard constraint) both nodes to be on the route of the vehicle X.

If your fleet is homogeneous (i.e. all vehicles are identical) then beside the hard/soft difference, the latter allow you to break some symmetry i.e.
the solver can't return the exact solution but with swapping vehicle X and vehicle Y routes since nodes are associated to a specific vehicle index.

**Reply by Mizux** [reply]
> when I have a time constraint (pickup and delivery windows) and I put a lock [...] What should I check or look into?

Since `ApplyLock` force nodes to be adjacent without any nodes in between you must verify that the chain is do able i.e.:
for the hypothetical chain A -> B you must have:
* There is a value in the Time Windows (TW) domain of A and B such that if vehicle visit A at `Ta` then it has the transit time to reach B at `Tb` with `Tb` being in TW domain of B.

Also verify vehicle has enough waiting time at A if needed e.g.
Suppose vehicle visit A at time `Ta = 10` (with TW at A being [10, 10])
also the transit is `Transit(A, B) = 10`
so vehicle can reach B from `Ta + Transit(A, B) = 10 + 10 = 20`
now suppose TW B is [30, 30] but SlackVar(A) is [0,5] then vehicle can wait at most 5 at A so we have `Ta + Transit(A, B) + Max_Wait(5) = 25` so vehicle can't reach B at 30 -> this lock `A->B` is impossible...

**Comment by odashevsky** [comment]
I think I found the problem why the solver does not sometimes find a solution when ApplyLocksToAllVehicles is used to lock in certain vehicle to certain locations. 

If I have locations A, B, C, and D and two vehicle X and Y and I use "ApplyLocksToAllVehicles" to force vehicle X to visit A and B and (!) technically A and B cannot be visited by vehicle X (too far and given pickup and drop off times constrains for A and B), the solver finds no solution. I suppose that's expected behavior, but my requirement is that I need to force A and B to be visited by X and I need the solver to accept that as a partial solution, what do I need to do? I cannot relax pickup and drop off times for A and B, the solver will generate unrealistic solution for other, unforced nodes. Please advise.

**Comment by jmarca** [comment]
I’m confused what you are asking. 

If you can’t make X visit A and B because it violates a constraint, and if you cannot relax the constraints, then you are stuck. 

I’d say relax constraints anyway, perhaps just for visits by vehicle X?  But you say it cannot be done. 

You’re the one who knows your problem the best, so if you say relaxing constraints is not possible, how are we supposed to know better?

James

**Reply by odashevsky** [reply]
I apologize, I see how what I wrote it might seem confusing. Let me clarify. 
Let's say there nodes with open timeframes: 
A (9:00 -9:05AM)
B (9:02 -9:07AM)
C (10 -11AM)
D (11 -12PM)
E (12 -1PM)

For simplicity sake, there's only one vehicle X. X cannot make A -> B given when they're open. But I can "force it" (business logic allows for that). The car then can easily make B->C -> D -> E (given that it 'magically" made it to B). So, I am locking "A->B" into the solution using ApplyLocksToAllVehicles. The solver comes back with "no solution found". I cannot relax windows for A or B, otherwise solver will create something like C->A->B->D->E, which is not valid. I was wondering if there's a way to make the solver use A->B and start with "A->B" (even if its unobtainable given other constraints in real world) and make it continue solving the rest of the solution.

**Reply by Mizux** [reply]
1. did you try to use the `routing.NextVar(A) = B` also you need to remove the TW for A/B (i.e. don't set it) otherwise as jmarca say your problem is infeasible from the solver point of view.

2. if to set your TW you're using `time_dim.CumulVar(X).SetRange(Y, Z)` then this is a hard constraint you could instead use:
```py
index = manager.NodeToIndex(X)
time_dim.SetCumulVarSoftLowerBound(index, Y-k1, penalty_per_unit_below_y)
time_dim.CumulVar(X).SetRange(Y, Z)
time_dim.SetCumulVarSoftUpperBound(index, Z+k2, penalty_per_unit_above_z)
 ```
aka solver can now visit node `index` in range `[Y-k1, Z+k2]` **but** will pay a penalty if not in the range `[Y, Z]`

**Comment by jmarca** [comment]
I’d try using a per-vehicle travel time callback, that allows vehicle X to get from A to B in 0 time. 

When “force it” is required, the relaxing you need is to teleport the vehicle. 

James

**Reply by odashevsky** [reply]
1. Tried that. No luck. With travel time set to 0 to A (from any location) to B (from any location) from A or B (to any location), if A and B open windows intersect - solver finds no solution.

2. Found another blocker: if in the scenario above, if I force, say, nodes **B and C,** (again, using ApplyLocksToAllVehicles), node A is ALWAYS dropped, even if there's plenty of time to make to B and C. But C->D->E is found by the solver.
If I lock C & D, A and B are always dropped, but D-> E is found!  In general, any node that should be visited before locked locations are always dropped.  Any idea why that is happening??

**Comment by jmarca** [comment]
Hmm, that isn't quite what I was recommending.  you said you set Any to A == 0, etc etc.  But I just said to set A to B = 0 for vehicle X, using 

AddDimensionWithVehicleTransitAndCapacity

So just vehicle X  can teleport from A to B, and only A to B gets zero travel time.

As to your other issue, I don't know.  Would be more helpful if you could generate a minimal reproducible problem code as a gist or whatever.

**Comment by odashevsky** [comment]
Guys, I am walking in circles here, I would really appreciate if someone could explain basic concepts for me. Namely:
1. What's the difference between `ReadAssignmentFromRoutes` and `RoutesToAssignment`? Both seem to load location matrices, why would one use over another?
2. Do I need to use routingModel.SolveFromAssignmentWithParameters or still stick to SolveWithParameters after I used either of the functions to load initial solution? Standard example at [(https://developers.google.com/optimization/routing/routing_tasks#setting-initial-routes-for-a-search)] doesn't really show why is it really even bothering dealing with initial solution except for to printing it. 
3. If I loaded initial solution with `ReadAssignmentFromRoutes` or `RoutesToAssignment`, will the solver "stick" to the initial solution and finish the solution? I am feeding an initial routes (assigned 25% of nodes to 50% of vehicles), but the solver completely ignores initial solution and generates another solution (granted, a better one, but I really need my initial solution there). Is that by intent?

I really would appreciate someone's input on this.

**Reply by Mizux** [reply]
https://gist.github.com/Mizux/0bd1daadc4e64130b62492c942f163d5#file-vrp_new_order-py-L139-L157 ?

**Reply by odashevsky** [reply]
Thanks for the link Mizux, I know you're trying to help. But the link does not explain if `SolveFromAssignmentWithParameters` (using initial solution from `ReadAssignmentFromRoutes `or `RoutesToAssignment) `"has to" use initial solution if its feasible. 
If these functions only use initial solution as 'suggestion" but feel free to not use it at all than I will stop pursuing this approach as it will not work for me.

**Reply by Mizux** [reply]
If I'm recall correctly, when using:
```py
initial_solution = routing.ReadAssignmentFromRoutes(initial_routes, True)
solution = routing.SolveFromAssignmentWithParameters(initial_solution, search_parameters)
```
you basically bypassing the first strategy heuristic so the local search will start from here. But unless some node are fixed (e.g. by using constraint on VehicleVar, NextVar, SetAllowedVehiclesForIndex) the solver is free to change any route etc. it is just a starting point.

**Comment by vuongdanghuy** [comment]
Hi @odashevsky 

With your 3rd question, if you don't adding more constraint, the solver will only use initial route as starting point to find better solution. If you want to keep your initial solution, I think you could do something like this:
```py
for v in vehicles:
    route = initialRoute[v]
    for node in route:
        routing.SetAllowedVehiclesForIndex([v], node)
```

It'll make initial route of a vehicle will only be visited by that vehicle alone. The solver will try to find better solution but still keep assigned node of a vehicle.

Hope this can help.

**Comment by odashevsky** [comment]
Everybody! Thank you so much for your replies. Based on every one's input I think I achieved a breakthrough. Again, thank you, I really appreciate everyone taking their time answering questions and providing suggestions!

**Reply by tesfayealex** [reply]
hello @odashevsky how were you able to solve your problem??

**Reply by odashevsky** [reply]
Yes, `SetAllowedVehiclesForIndex` did it for me
